include/revng/Model/Types.h

```diff
@@ -0,0 +1,335 @@
+  EnumType &operator=(EnumType &&) = default;
+
+  static bool classof(const Type *T) { return T->getKind() == TK_Enum; }
+
+  // Kind must be TK_Enum; Underlying type must verify and be an integral
+  // primitive type; there must be at leas one entry, and all the entries must
+  // verify.
+  bool verify() const debug_function;
+};
+
+struct TypedefType : public Type {
+  QualifiedType UnderlyingType;
+
+  TypedefType() = delete;
+
+  TypedefType(llvm::StringRef NameRef) :
+    Type(NameRef, TK_Typedef), UnderlyingType() {}
+
+  TypedefType(const TypedefType &) = default;
+  TypedefType &operator=(const TypedefType &) = default;
+
+  TypedefType(TypedefType &&) = default;
+  TypedefType &operator=(TypedefType &&) = default;
+
+  static bool classof(const Type *T) { return T->getKind() == TK_Typedef; }
+
+  // Kind must be TK_Typedef and Underlying type must verify
+  bool verify() const debug_function;
+};
+
+struct AggregateField {
+  std::string Name;
+  QualifiedType FieldType;
+
+  AggregateField() = delete;
+  ~AggregateField() = default;
+
+  AggregateField(llvm::StringRef NameRef) : Name(NameRef), FieldType() {}
+
+  AggregateField(const AggregateField &) = default;
+  AggregateField &operator=(const AggregateField &) = default;
+
+  AggregateField(AggregateField &&) = default;
+  AggregateField &operator=(AggregateField &&) = default;
+
+public:
+  bool verify() const { return not Name.empty() and FieldType.verify(); }
+};
+
+} // end namespace flatc
+
+template<>
+struct KeyedObjectTraits<flatc::AggregateField>
+  : public NamedKeyedObjectTraits<flatc::AggregateField> {};
+
+namespace flatc {
+
+struct StructType : public Type {
+  std::vector<AggregateField> Fields;
+
+  StructType() = delete;
+
+  StructType(llvm::StringRef NameRef) : Type(NameRef, TK_Struct), Fields() {}
+
+  StructType(const StructType &) = default;
+  StructType &operator=(const StructType &) = default;
+
+  StructType(StructType &&) = default;
+  StructType &operator=(StructType &&) = default;
+
+  static bool classof(const Type *T) { return T->getKind() == TK_Struct; }
+
+  // Kind must be TK_Struct, Fields must not be empty, all Fields must verify
+  // and Fields must not have duplicate names.
+  bool verify() const debug_function;
+};
+
+struct UnionType : public Type {
+  SortedVector<AggregateField> Fields;
+
+  UnionType() = delete;
+
+  UnionType(llvm::StringRef NameRef) : Type(NameRef, TK_Union), Fields() {}
+
+  UnionType(const UnionType &) = default;
+  UnionType &operator=(const UnionType &) = default;
+
+  UnionType(UnionType &&) = default;
+  UnionType &operator=(UnionType &&) = default;
+
+  static bool classof(const Type *T) { return T->getKind() == TK_Union; }
+
+  // Kind must be TK_Union, Fields must not be empty, all Fields must verify
+  bool verify() const debug_function;
+};
+
+struct FunctionPtrType : public Type {
+  QualifiedType ReturnType;
+  std::vector<QualifiedType> ArgumentTypes;
+  bool IsVariadic;
```

Comment:
Let's not introduce expressive power we don't have short-term plans to handle.

Reply:
So, you want me to drop `IsVariadic` alltogether right now?

I'm not sure we don't need `IsVariadic` already. For instance, last time I checked `newpc` was variadic, and `GeneratedIRReference.rst` still says it is. Isn't that anymore the case? Or are you saying that you never want to describe `newpc` in the model?

And if we never want to describe it in the model, we are also implying that we cannot emit it in C, since `flatc` is meant to eventually become the only way we use to describe function types. Are you ok with this?

Reply:
`newpc` is not a lifted function so it doesn't need to be represented in the model.

I'm just saying, we have really foggy plans for variadic functions. Let's just postpone this, otherwise we'll have a field no one is using. Just noise.